### PR TITLE
Update LUFA to newest (18-07-2021) and add QMK-HID Bootloader support

### PR DESCRIPTION
**Pull request by drashna**
## Description

Like the title says. 
Updates LUFA to the latest version, as well as adds support for generating the lufa HID bootloader, as well as flashing it.

## Types of Changes

- [x] Core
- [x] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: bootloader.mk [hunk 1]
```diff
@@ -67,6 +67,11 @@ ifeq ($(strip $(BOOTLOADER)), qmk-dfu)
         BOOTLOADER_SIZE = 8192
     endif
 endif
+ifeq ($(strip $(BOOTLOADER)), qmk-hid)
```
Comment (fauxpark) [inline bootloader.mk:70]: Also add to list at the top.

Comment (drashna) [reply bootloader.mk:71]: Added

**Comment by drashna** [comment]
Added additional documentation. 

Also want to add a split left/right target, but doesn't support that natively.  Would have to be a hack like dfu-util.  teensy/halfkay could use that too.

File: docs/flashing.md [hunk 2]
```diff
@@ -171,6 +171,55 @@ Flashing sequence:
...
+
+The manufacturer and product strings are automatically pulled from `config.h`, with " Bootloader" appended to the product string.
+
+To generate this bootloader, use the `bootloader` target, eg. `make planck/rev4:default:bootloader`. To generate a production-ready .hex file (combining QMK and the bootloader), use the `production` target, eg. `make planck/rev4:default:production`.
+
+
+Compatible flashers:
+
+* TBD
+  * Currently, you need to either use the [python script](https://github.com/qmk/lufa/tree/master/Bootloaders/HID/HostLoaderApp_python) in the lufa repo, or compile the [`hid_bootloader_cli`](https://github.com/qmk/lufa/tree/master/Bootloaders/HID/HostLoaderApp) from the LUFA repo.   Homebrew may (will) have support for this directly (via `brew install qmk/qmk/hid_bootloader_cli`)
```
Comment (fauxpark) [inline docs/flashing.md:207]:
```suggestion
  * Currently, you need to either use the [Python script](https://github.com/qmk/lufa/tree/master/Bootloaders/HID/HostLoaderApp_python), or compile [`hid_bootloader_cli`](https://github.com/qmk/lufa/tree/master/Bootloaders/HID/HostLoaderApp), from the LUFA repo. Homebrew may (will) have support for this directly (via `brew install qmk/qmk/hid_bootloader_cli`).
```

File: docs/flashing.md [hunk 3]
```diff
@@ -171,6 +171,55 @@ Flashing sequence:
...
+//#define QMK_SPEAKER C6
+```
+
+Currently we do not recommend making `QMK_ESC` the same key as the one designated for [Bootmagic Lite](feature_bootmagic.md#bootmagic-lite), as holding it down will cause the MCU to loop back and forth between entering and exiting the bootloader.
+
+The manufacturer and product strings are automatically pulled from `config.h`, with " Bootloader" appended to the product string.
+
+To generate this bootloader, use the `bootloader` target, eg. `make planck/rev4:default:bootloader`. To generate a production-ready .hex file (combining QMK and the bootloader), use the `production` target, eg. `make planck/rev4:default:production`.
+
+
```
Comment (fauxpark) [inline docs/flashing.md:203]:
```suggestion
```

File: docs/flashing.md [hunk 4]
```diff
@@ -171,6 +171,55 @@ Flashing sequence:
 3. Flash a .hex file
 4. Reset the device into application mode (may be done automatically)
 
+
+### QMK HID
+
+QMK maintains [a fork of the LUFA HID bootloader](https://github.com/qmk/lufa/tree/master/Bootloaders/HID), which uses a USB HID Endpoint for flashing in the way that the PJRC's Teensy Loader flasher and Halfkay bootloader work. Additionally, it performs a simple matrix scan for exiting the bootloader and returning to the application, as well as flashing an LED/making a ticking noise with a speaker when things are happening. 
+
+
```
Comment (fauxpark) [inline docs/flashing.md:179]:
```suggestion
```

File: docs/flashing.md [hunk 5]
```diff
@@ -171,6 +171,55 @@ Flashing sequence:
 3. Flash a .hex file
 4. Reset the device into application mode (may be done automatically)
 
+
```
Comment (fauxpark) [inline docs/flashing.md:174]:
```suggestion
```

File: docs/flashing.md [hunk 6]
```diff
@@ -171,6 +171,55 @@ Flashing sequence:
...
+### QMK HID
+
+QMK maintains [a fork of the LUFA HID bootloader](https://github.com/qmk/lufa/tree/master/Bootloaders/HID), which uses a USB HID Endpoint for flashing in the way that the PJRC's Teensy Loader flasher and Halfkay bootloader work. Additionally, it performs a simple matrix scan for exiting the bootloader and returning to the application, as well as flashing an LED/making a ticking noise with a speaker when things are happening.
```
Comment (fauxpark) [inline docs/flashing.md:177]:
```suggestion
QMK maintains [a fork of the LUFA HID bootloader](https://github.com/qmk/lufa/tree/master/Bootloaders/HID), which uses a USB HID Endpoint for flashing in the way that the PJRC's Teensy Loader flasher and HalfKay bootloader work. Additionally, it performs a simple matrix scan for exiting the bootloader and returning to the application, as well as flashing an LED/making a ticking noise with a speaker when things are happening.
```

File: docs/flashing.md [hunk 7]
```diff
@@ -171,6 +171,55 @@ Flashing sequence:
...
+
+
+Compatible flashers:
+
+* TBD
+  * Currently, you need to either use the [python script](https://github.com/qmk/lufa/tree/master/Bootloaders/HID/HostLoaderApp_python) in the lufa repo, or compile the [`hid_bootloader_cli`](https://github.com/qmk/lufa/tree/master/Bootloaders/HID/HostLoaderApp) from the LUFA repo.   Homebrew may (will) have support for this directly (via `brew install qmk/qmk/hid_bootloader_cli`)
+
+Flashing sequence:
+
+1. Enter the bootloader using any of the following methods (you only have 7 seconds to flash once it enters):
```
Comment (fauxpark) [inline docs/flashing.md:211]:
```suggestion
1. Enter the bootloader using any of the following methods:
```
I guess this is copied from the HalfKay section - only Caterina has this 7-second timeout, so it should be removed from there too.

File: docs/flashing.md [hunk 8]
```diff
@@ -171,6 +171,55 @@ Flashing sequence:
...
+    * Press the `RESET` keycode
+    * Press the `RESET` button on the Teensy or PCB if available
```
Comment (fauxpark) [inline docs/flashing.md:213]:
```suggestion
    * Press the `RESET` button on the PCB if available
```

File: tmk_core/avr.mk [hunk 9]
```diff
@@ -295,19 +305,28 @@ extcoff: $(BUILD_DIR)/$(TARGET).elf
 	@$(SECHO) $(MSG_EXTENDED_COFF) $(BUILD_DIR)/$(TARGET).cof
 	$(COFFCONVERT) -O coff-ext-avr $< $(BUILD_DIR)/$(TARGET).cof
 
-bootloader:
-ifneq ($(strip $(BOOTLOADER)), qmk-dfu)
-	$(error Please set BOOTLOADER = qmk-dfu first!)
+
+
```
Comment (fauxpark) [inline tmk_core/avr.mk:309]:
```suggestion
```

File: tmk_core/avr.mk [hunk 10]
```diff
@@ -237,6 +237,16 @@ endef
 bootloadHID: $(BUILD_DIR)/$(TARGET).hex check-size cpfirmware
 	$(call EXEC_BOOTLOADHID)
 
+
```
Comment (fauxpark) [inline tmk_core/avr.mk:240]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — CHANGES_REQUESTED** [review]
Should also add the following to the table at the end of `driver_installation_zadig.md`:

```
|`qmk-hid`    |(keyboard name) Bootloader    |`03EB:2067`   |HidUsb |
```

**Review by fauxpark — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Pull request merged**